fem/restriction.cpp

```diff
@@ -594,119 +595,246 @@ void L2ElementRestriction::FillJAndData(const Vector &ea_data,
-                  }
-               }
-               break;
-            case 1: // SOUTH
-               for (int i = 0; i < dof1d; ++i)
-               {
-                  for (int j = 0; j < dof1d; ++j)
-                  {
-                     face_map[i+j*dof1d] = i + j*dof1d*dof1d;
-                  }
-               }
-               break;
-            case 2: // EAST
-               for (int i = 0; i < dof1d; ++i)
-               {
-                  for (int j = 0; j < dof1d; ++j)
-                  {
-                     face_map[i+j*dof1d] = dof1d-1 + i*dof1d + j*dof1d*dof1d;
-                  }
-               }
-               break;
-            case 3: // NORTH
-               for (int i = 0; i < dof1d; ++i)
-               {
-                  for (int j = 0; j < dof1d; ++j)
-                  {
-                     face_map[i+j*dof1d] = (dof1d-1)*dof1d + i + j*dof1d*dof1d;
-                  }
-               }
-               break;
-            case 4: // WEST
-               for (int i = 0; i < dof1d; ++i)
-               {
-                  for (int j = 0; j < dof1d; ++j)
-                  {
-                     face_map[i+j*dof1d] = i*dof1d + j*dof1d*dof1d;
-                  }
-               }
-               break;
-            case 5: // TOP
-               for (int i = 0; i < dof1d; ++i)
-               {
-                  for (int j = 0; j < dof1d; ++j)
-                  {
-                     face_map[i+j*dof1d] = (dof1d-1)*dof1d*dof1d + i + j*dof1d;
-                  }
-               }
-               break;
+            case 0: offsets = {p*pp1}; break; // y = 0
+            case 1: offsets = {pp1 - 1}; break; // x = 1
+            case 2: offsets = {p*pp1 + p*(pp1 - 1)}; break; // y = 1
+            case 3: offsets = {0}; break; // x = 0
+         }
+         break;
+      case 3:
+      {
+         const int n_dof_per_dim = p*p*pp1;
+         const auto f = GetFaceNormal3D(face_id);
+         const int face_normal = f.first, level = f.second;
+         if (face_normal == 0) // x-normal
+         {
+            offsets = {level ? pp1 - 1 : 0};
+            strides = {pp1, p*pp1};
+         }
+         else if (face_normal == 1) // y-normal
+         {
+            offsets = {n_dof_per_dim + (level ? p*(pp1 - 1) : 0)};
+            strides = {1, p*pp1};
+         }
+         else if (face_normal == 2) // z-normal
+         {
+            offsets = {2*n_dof_per_dim + (level ? p*p*(pp1 - 1) : 0)};
+            strides = {1, p};
          }
          break;
+      }
    }
+   std::vector<int> n_dofs(dim - 1, p);
+   FillFaceMap(n_face_dofs, offsets, strides, n_dofs, face_map);
 }
 
-H1FaceRestriction::H1FaceRestriction(const FiniteElementSpace &fes,
-                                     const ElementDofOrdering e_ordering,
-                                     const FaceType type,
-                                     bool build)
+void GetFaceDofs(const FiniteElementSpace &fes,
```

Comment:
Wouldn't it make more sense to make this function a `virtual` method of `FiniteElementCollection`? This kind of `dynamic_cast` is usually strongly discouraged.

Reply:
I think that's a good idea — let me see if I can make that change. It may also partially address your concerns below about having a monolithic class/maintainability.

Reply:
This is what it looks like: https://github.com/mfem/mfem/compare/rt-face-restriction...fe-face-map

(It didn't make sense to make it part of the FE collection, these are really specific to individual finite elements).

The good thing is that instead of switch statements, we can use the inheritance hierarchy to dispatch to the correct implementation.

The downside is that the implementations are kind of scattered across the different `fe` files.

Let me know what you think.

Reply:
@YohannDudouit did you get a chance to look at the proposed change? Which version do you prefer?

Reply:
Yes, I do think it is an improvement. My only issue is that in the general case `GetFaceMap` does not make sense, it only makes sense for "closed" basis, but I don't have a good solution for that (traits? 😋).

Reply:
OK, I merged it into this branch.

Yeah, good point about not making sense for every element/basis type, but I think there are enough cases of `MFEM_ABORT("method is not implemented for this element");` elsewhere that this won't stick out.